core-bundle/src/Image/Studio/Studio.php

```diff
@@ -0,0 +1,622 @@
+
+    /**
+     * Check if the provided file uri has a valid image file extension.
+     */
+    private function hasValidImageType(?string $uri): ?bool
+    {
+        if (null === $uri) {
+            return null;
+        }
+
+        /** @var Config $configAdapter */
+        $configAdapter = $this->getAdapter(Config::class);
+
+        $validImageTypes = explode(',', $configAdapter->get('validImageTypes'));
+
+        return \in_array(Path::getExtension($uri), $validImageTypes, true);
+    }
+
+    private function isExternalUrl(?string $uri): ?bool
+    {
+        if (null === $uri) {
+            return null;
+        }
+
+        return preg_match('#^https?://#', $uri);
+    }
+
+    /**
+     * Try to get a light box size configuration from the current page's layout
+     * model. Will return null if not defined or not in a request context.
+     */
+    private function getLightBoxSizeConfiguration(): ?array
+    {
+        $request = $this->getRequest();
+
+        if (null === $request || !$request->attributes->has('pageModel')) {
+            return null;
+        }
+
+        /** @var PageModel $pageAdapter */
+        $pageAdapter = $this->getAdapter(PageModel::class);
+
+        /** @var PageModel|null $page */
+        $page = $pageAdapter->findByPk($request->attributes->get('pageModel'));
+
+        if (null === $page || null === $page->layout) {
+            return null;
+        }
+
+        /** @var LayoutModel $layoutModelAdapter */
+        $layoutModelAdapter = $this->getAdapter(LayoutModel::class);
+
+        /** @var LayoutModel|null $layoutModel */
+        $layoutModel = $layoutModelAdapter->findByPk($page->layout);
+
+        if (null === $layoutModel || empty($layoutModel->lightboxSize)) {
+            return null;
+        }
+
+        return StringUtil::deserialize($layoutModel->lightboxSize, true);
+    }
+
+    /**
+     * Try to locate a file by querying the DBAFS ($identifier = uuid/id/path),
+     * fallback to interpret $identifier as absolute/relative file path.
+     */
+    private function locateResource($identifier, FilesModel &$filesModel = null): string
+    {
+        $dbafsItem = true;
+
+        if ($identifier instanceof FilesModel) {
+            $filesModel = $identifier;
+        } else {
+            // Try to interpret the identifier as a DBAFS item (and get a file model reference)
+
+            /** @var Validator $validatorAdapter */
+            $validatorAdapter = $this->getAdapter(Validator::class);
+
+            /** @var FilesModel $filesModelAdapter */
+            $filesModelAdapter = $this->getAdapter(FilesModel::class);
+
+            if ($validatorAdapter->isUuid($identifier)) {
+                $filesModel = $filesModelAdapter->findByUuid($identifier);
+            } elseif (is_numeric($identifier)) {
+                $filesModel = $filesModelAdapter->findById((int) $identifier);
+            } else {
+                $filesModel = $filesModelAdapter->findByPath($identifier);
```

Comment:
I think we should query the database only if the path starts with the upload directory, see https://github.com/contao/contao/blob/b4460960d6d7dc35bf4a87bcb7a503bce1df0d6d/core-bundle/src/Image/ImageFactory.php#L305-L307 for performance reasons.

Reply:
:+1: Changed now.  I'm not sure if we should also support absolute paths to DBAFS resources. Right now they are treated like every other absolute path.